### ISSUE TITLE
Error building kaldi-nnet3 in Visual Studio 2015: error C2664: 'std::pair<kaldi::int32,kaldi::int32> std::make_pair<T,T>(_Ty1 &&,_Ty2 &&)': cannot convert argument 1 from 'kaldi::int32' to 'kaldi::int32 &&'

### ISSUE THREAD

**Issue by ceefour** [kaldi-asr/kaldi#596]
Environment: Visual Studio Community 2015 Update 1 on Windows 10 Pro 64-bit, Intel MKL.

```
19>------ Build started: Project: kaldi-nnet3, Configuration: Debug x64 ------
19>  nnet-utils.cc
19>  nnet-training.cc
19>  nnet-test-utils.cc
19>  nnet-simple-component.cc
19>  nnet-parse.cc
19>  nnet-optimize.cc
19>  nnet-optimize-utils.cc
19>  nnet-nnet.cc
19>  nnet-graph.cc
19>  nnet-general-component.cc
19>D:\kaldi-trunk\src\base/io-funcs-inl.h(185): error C2664: 'std::pair<kaldi::int32,kaldi::int32> std::make_pair<T,T>(_Ty1 &&,_Ty2 &&)': cannot convert argument 1 from 'kaldi::int32' to 'kaldi::int32 &&'
19>          with
19>          [
19>              T=kaldi::int32,
19>              _Ty1=kaldi::int32,
19>              _Ty2=kaldi::int32
19>          ]
19>  D:\kaldi-trunk\src\base/io-funcs-inl.h(185): note: You cannot bind an lvalue to an rvalue reference
19>  ..\..\..\src\nnet3\nnet-general-component.cc(96): note: see reference to function template instantiation 'void kaldi::ReadIntegerPairVector<kaldi::int32>(std::istream &,bool,std::vector<std::pair<kaldi::int32,kaldi::int32>,std::allocator<_Ty>> *)' being compiled
19>          with
19>          [
19>              _Ty=std::pair<kaldi::int32,kaldi::int32>
19>          ]
19>  nnet-example.cc
19>  nnet-example-utils.cc
19>  nnet-discriminative-training.cc
19>  nnet-discriminative-example.cc
19>  nnet-discriminative-diagnostics.cc
19>  nnet-diagnostics.cc
19>  nnet-descriptor.cc
19>  nnet-compute.cc
19>  nnet-computation.cc
19>D:\kaldi-trunk\src\base/io-funcs-inl.h(185): error C2664: 'std::pair<kaldi::int32,kaldi::int32> std::make_pair<T,T>(_Ty1 &&,_Ty2 &&)': cannot convert argument 1 from 'kaldi::int32' to 'kaldi::int32 &&'
19>          with
19>          [
19>              T=kaldi::int32,
19>              _Ty1=kaldi::int32,
19>              _Ty2=kaldi::int32
19>          ]
19>  D:\kaldi-trunk\src\base/io-funcs-inl.h(185): note: You cannot bind an lvalue to an rvalue reference
19>  ..\..\..\src\nnet3\nnet-computation.cc(720): note: see reference to function template instantiation 'void kaldi::ReadIntegerPairVector<kaldi::int32>(std::istream &,bool,std::vector<std::pair<kaldi::int32,kaldi::int32>,std::allocator<_Ty>> *)' being compiled
19>          with
19>          [
19>              _Ty=std::pair<kaldi::int32,kaldi::int32>
19>          ]
19>  nnet-computation-graph.cc
19>  Generating Code...
19>  Compiling...
19>  nnet-component-itf.cc
19>  nnet-compile.cc
19>..\..\..\src\nnet3\nnet-compile.cc(224): warning C4390: ';': empty controlled statement found; is this the intent?
19>  nnet-compile-utils.cc
19>  nnet-common.cc
19>  nnet-combine.cc
19>  nnet-chain-training.cc
19>  nnet-chain-example.cc
19>  nnet-chain-diagnostics.cc
19>  nnet-chain-combine.cc
19>  nnet-analyze.cc
19>  nnet-am-decodable-simple.cc
19>  natural-gradient-online.cc
19>  discriminative-training.cc
19>..\..\..\src\nnet3\discriminative-training.cc(633): error C2146: syntax error: missing ')' before identifier 'and'
19>..\..\..\src\nnet3\discriminative-training.cc(633): error C2065: 'and': undeclared identifier
19>..\..\..\src\nnet3\discriminative-training.cc(633): error C2146: syntax error: missing ';' before identifier 'pdf_id'
19>..\..\..\src\nnet3\discriminative-training.cc(633): error C2059: syntax error: ')'
19>..\..\..\src\nnet3\discriminative-training.cc(637): error C2059: syntax error: ';'
19>..\..\..\src\nnet3\discriminative-training.cc(644): error C2059: syntax error: '}'
19>..\..\..\src\nnet3\discriminative-training.cc(644): error C2143: syntax error: missing ';' before '}'
19>  discriminative-supervision.cc
19>  am-nnet-simple.cc
19>  Generating Code...
 ```

Potentially useful resources:

* [Breaking Changes in Visual C++ 2015](https://msdn.microsoft.com/en-us/library/bb531344.aspx)
* [c++ - Breaking change in C++11 with make_pair(_Ty1&& _Val1, const _Ty2& _Val2) - Stack Overflow](http://stackoverflow.com/questions/14623958/breaking-change-in-c11-with-make-pair-ty1-val1-const-ty2-val2)
* [c2664 in Visual Studio 2012 when using make_pair - Stack Overflow](http://stackoverflow.com/questions/13902348/c2664-in-visual-studio-2012-when-using-make-pair)

Original thread: https://groups.google.com/d/msg/kaldi-help/GSOr3AWnuWM/yZ7-N79QBgAJ

**Comment by jtrmal** [comment]
I tried to compile it and it would be quite laborous to fix the compilation issues.
It seems to me that the type-lookup changed again so anytime there is combination of both namespaces fst and kaldi, the lookup won't work (it will produce errors such as "uint32: ambiguous symbol")
I will ask @danpovey if he will see any solution but unless there is something less labor-intensive, I'm not planning for supporting VS2015  ATM, you will have to use VS2013.

**Comment by ceefour** [comment]
Can you merge #598? At least there'll be one less error to fix :D

**Comment by danpovey** [comment]
@xiaohui-zhang, can you please grep everywhere in the Kaldi source for
make_pair and convert it to use the constructor?
This can't be done by a script, as you need to know the type.
The error in discriminative-training.cc was an 'and' that should have been
'&&'. I just fixed it. @vimalmanohar, take note!

There are some compilation problems that can be fixed by statements like
typedef kaldi::int32 int32;
if you are in the global namespace.
this is often needed in main() functions.

Dan

**Comment by jtrmal** [comment]
There were many more issues with what I assume was Vimal's code (for
example mistakenly used 'and' instead of '&&' -- that compiles under gcc as
the C++ standard defines 'trigraphs', however, trigraphs are removed in
C++17. And because VC++ is inconsistent and implements whatever it feels
like (sometimes their strategy feels like a random walk), they removed
trigraphs even now, so the code does not compile.

the typedef actually does not seem to solve the issues, I tried. I'll show
you an example later.
y.

**Comment by xiaohui-zhang** [comment]
sure. will do soon.

On Thu, Mar 17, 2016 at 2:52 PM, Daniel Povey <notifications@github.com>
wrote:

> @xiaohui-zhang, can you please grep everywhere in the Kaldi source for
> make_pair and convert it to use the constructor?
> This can't be done by a script, as you need to know the type.
> The error in discriminative-training.cc was an 'and' that should have been
> '&&'. I just fixed it. @vimalmanohar, take note!
>
> There are some compilation problems that can be fixed by statements like
> typedef kaldi::int32 int32;
> if you are in the global namespace.
> this is often needed in main() functions.
>
> Dan
>
>
>
> On Thu, Mar 17, 2016 at 11:29 AM, Hendy Irawan <notifications@github.com>
> wrote:
>
> > Can you merge #598 <https://github.com/kaldi-asr/kaldi/pull/598>? At
> > least there'll be one less error to fix :D
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/issues/596#issuecomment-197932730>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/issues/596#issuecomment-198032552>
>



-- 
Xiao-hui

**Comment by jtrmal** [comment]
Guys, let's not fragment the effort. I have most of the other fixes ready.
Doing another round of merge is not a necessary hassle.
y.

**Comment by danpovey** [comment]
I know in the past we have had namespace lookup issues in VS that we were
not able to fix, because when templates are instantiated it makes names
available from the namespace in which the template as instantiated in.
This is totally against the C++ standard and there's really no way to
reconcile it.
Dan

**Comment by kkm000** [comment]
> trigraphs are removed in C++17. And because VC++ is inconsistent and implements whatever it feels like

As is any other compiler. Sigh...

**Comment by jtrmal** [comment]
Note: this would be best to read in github www, not the e-mail:
I have a code (lattice-compose.cc) that starts with
```
int main(int argc, char *argv[]) {
  try {
    using namespace kaldi;
    //typedef kaldi::int32 int32;
    //typedef kaldi::int64 int64;
    using fst::SymbolTable;
    using fst::VectorFst;
    using fst::StdArc;
```
then later in the code there is a statement
```
      fst::StdToLatticeMapper<BaseFloat> mapper;
```
which generates errors like this
```
1>c:\users\yenda\downloads\kaldi\src\latbin\lattice-compose.cc(49): error C2872: 'int32': ambiguous symbol
1>  c:\users\yenda\downloads\kaldi\tools\openfstwin-1.3.3\src\include\fst\types.h(30): note: could be 'int32_t int32'
1>  c:\users\yenda\downloads\kaldi\src\base\kaldi-types.h(47): note: or       'kaldi::int32'
1>c:\users\yenda\downloads\kaldi\src\latbin\lattice-compose.cc(50): error C2872: 'int32': ambiguous symbol
1>  c:\users\yenda\downloads\kaldi\tools\openfstwin-1.3.3\src\include\fst\types.h(30): note: could be 'int32_t int32'
1>  c:\users\yenda\downloads\kaldi\src\base\kaldi-types.h(47): note: or       'kaldi::int32'
1>c:\users\yenda\downloads\kaldi\src\latbin\lattice-compose.cc(67): error C2872: 'int32': ambiguous symbol
1>  c:\users\yenda\downloads\kaldi\tools\openfstwin-1.3.3\src\include\fst\types.h(30): note: could be 'int32_t int32'
1>  c:\users\yenda\downloads\kaldi\src\base\kaldi-types.h(47): note: or       'kaldi::int32'
1>c:\users\yenda\downloads\kaldi\src\fstext\lattice-utils.h(218): error C2872: 'uint64': ambiguous symbol
1>  c:\users\yenda\downloads\kaldi\tools\openfstwin-1.3.3\src\include\fst\types.h(36): note: could be 'uint64_t uint64'
1>  c:\users\yenda\downloads\kaldi\src\base\kaldi-types.h(45): note: or       'kaldi::uint64'
1>  c:\users\yenda\downloads\kaldi\src\latbin\lattice-compose.cc(89): note: see reference to class template instantiation 'fst::StdToLatticeMapper<kaldi::BaseFloat>' being compiled
```

so far, I didn't find any other solution than avoid ``import kaldi`` and then prepend ``kaldi::`` (or use ``using identifier`` for each member of kaldi namespace used in the code)
y.

**Comment by jtrmal** [comment]
@kkm000 , any ideas?

**Comment by jtrmal** [comment]
I can kind of understand the problem -- calling ``using kaldi`` imports the kaldi types into the global namespace, where are already types of the same name defined by openfst (which defines the types globally instead of defining it in fst::), but as the typedef trick is not helping, I dont know how to help the compiler to resolve this. This affects even implicit template parameters and the errors are sometimes extremely hard to trace.

**Comment by kkm000** [comment]
That's almost certainly the same VC template expansion scoping bug Dan has mentioned. 

I am not set up to compile at this very moment (need to pull latest code etc), here's a workaround idea I would try the first: `typedef fst::StdToLatticeMapper<kaldi::BaseFloat> something;` before importing the namespace.

**Comment by jtrmal** [comment]
So you aim at enforcing the type deduction before importing the whole
namespace?
Anyway, thanks, I'll try something along these lines;
y.

On Thu, Mar 17, 2016 at 4:04 PM, Kirill Katsnelson <notifications@github.com

**Comment by kkm000** [comment]
Yes, that's the idea.

**Comment by jtrmal** [comment]
resolved by #600
